### DISCUSSION TITLE
How the Panels are linked with Queries

### DISCUSSION THREAD

**Discussion by celian-garcia** [perses/perses#1922] [Design Discussions]
Here is a summary of a problem that I exposed recently in sync weekly meeting. 

### Context
Currently we have several type of plugins (here in red) that a plugin can extend (here in blue).
> [!CAUTION]
> What is important to note is that
> - **red** definitions are located into the **perses** codebase 
> - **blue** definitions  are located into the **plugin** codebase

![image](https://github.com/perses/perses/assets/10258608/838c377b-29ec-4285-a5ca-485afc6be5dd)

### The problem
If somebody want to create some new plugin of type "Panel" they might struggle to subscribe into the different existing query types existing. They will probably have to create a new query type and need to create a PR to Perses.
This is basically what happened when RedHat introduced Tracing.

### The solutions
Basically looking at the code, what I realized is that `TimeSeriesQuery` and `TraceQuery` are just here to establish a contract. A data format that needs to be respected by the plugins implementing it. 

Two solutions foreseen so far. 
1/ Introduce a third layer data, that would defined a data format. And the Query/Panel plugins should declares what format they return/accept.
![image](https://github.com/perses/perses/assets/10258608/15f3aa3f-b1bb-4ed5-b673-96ce3f6287f8)

2/ There is also the solution of having a common and sufficiently generic data format like DataFrame that would be used by each Panel and Query. 
But then it would require more things into the Panel implementation. (e.g for timeserieschart: where is the located the timestamp column) 
![image](https://github.com/perses/perses/assets/10258608/96c3c2b1-8656-48f0-8996-1b97f387265b)

**Comment by AntoineThebaud** [comment]
We discussed this topic again with @Nexucis and on our side we are actually more in favor of keeping the things as they are, aka have the data formats "contracts" as part of the Perses repo / core. This way any plugin could easily import them & it would encourage plugins to converge towards / reuse standard formats, which is likely to increase the number of synergies between panel & query plugins developed by unrelated developers.
In case someone needs a new data format to be added because we never encountered its observability use case before (e.g the GeoMap mentioned above), (s)he would open a PR or an issue to the Perses repo, and we would help to refine the proposal towards a standard/generic/future-proof data model. This approach can look rigid / not plugin-friendly, but this entry cost would actually help to increase the plugin's "compatibility matrix" thus the overall modularity in the end, so worth it IMHO.
I think the difficulties that RedHat encountered with Tracing are to be imputed to the fact that they were "pilots despite themselves" and uncovered this activity for which we had no clear strategy in place.

**Reply by celian-garcia** [reply]
First solution describes actually something where people can declare their own data model as a plugin. And other plugins could reuse it freely integrating it in their dependencies. 

What you describe is the current architecture.

**Reply by AntoineThebaud** [reply]
> First solution describes actually something where people can declare their own data model as a plugin. And other plugins could reuse it freely integrating it in their dependencies.
> 
> What you describe is the current architecture.

Indeed my bad! I updated my post above accordingly.

**Reply by celian-garcia** [reply]
Okay yes. At the end I kinda agree with you that this part as been difficult for RedHat as they were the first having to create the traces data model. This would be the same for the first that will implement LogQuery. We can maybe propose a draft data model already for logging as I believe it's very close to timeseries but with a string value.

**Reply by jgbernalp** [reply]
I'd say the contract could be the DataFrame, meaning the standard format. Which already increases the synergy between panel and queries because every query can be bound to every panel, so there wont be an entry cost for being compatible.

**Reply by AntoineThebaud** [reply]
I fear this would lead to mismatching types between panels & queries that could have been compatible. E.g a given query produces data frames with 2 dimensions "time" and "value" while a given panel expects data frames with 2 dimensions "timestamp" & "val".

**Reply by celian-garcia** [reply]
Yes I think that with dataframe there would be an implicit contract that needs to be implemented like timestamps are always in "timestamp" field value is always in "value" field. 

Also in the plugin discussion #2085 @andreasgerstmayr noted the presence of interesting rich fields that could be hard to type in a generic format.

I personally feel like we could have the data frame type alongside the existing type. I imagine Timeseries Query can return TimeseriesData and Dataframe. That would be useful for generic panels like Table panel

**Reply by celian-garcia** [reply]
That could take the shape of inheritance between TimeseriesData and Dataframe. ``TimeSeriesData implements/extends Dataframe`` with some generic method to get data for generic panels, and some extra methods for specialized panels

**Reply by jgbernalp** [reply]
yes I agree with @celian-garcia, a TimeSeriesData is just a DataFrame with an extended contract. And for more complex data shapes such as tracing the same will apply.

**Reply by MitchellJThomas** [reply]
Pardon me as I'm new here, but I was wondering if CueLang's [concept of values and types](https://cuelang.org/docs/tour/types/types/) might be useful in the context of describing the shape of data flowing between a Panel and a Query.

**Reply by celian-garcia** [reply]
Here we speak about generic OOP concepts to describe the format of the data that will be passed from query to panels. This will be exclusively managed with typescript in the UI only. Nothing saved in the backend with that format so cuelang is unrelated. Or am I missing your point?

**Reply by MitchellJThomas** [reply]
Ah, my apologies I didn't realize the conversation was focused on the browser side of things.
